### PR TITLE
Heuristic update

### PR DESCRIPTION
**Pull request by rspeer**
This adjusts some heuristics that caused false positives when using common punctuation symbols from the Windows-1252 set, by expanding the `ENDING_PUNCT_RE` to the more general `COMMON_SYMBOL_RE`. This generalizes the old `punctuation-exception` branch. And with the current heuristics, it appears to be fine to include the ACUTE ACCENT in there, fixing the Hogwarts false positive.

It puts math and currency symbols into the same character class, because it's actually not that weird for them to appear next to each other.

A few versions ago, we set aside a range of codepoints that we'd assume were emoji. There are so many new emoji that they've already exceeded that range. So now the range is larger.

### CODE REVIEW

File: tests/test_characters.py [hunk 1]
```diff
@@ -25,6 +25,8 @@ def test_bmp_characters():
                     fixed, plan = fix_encoding_and_explain(garb)
                     eq_(fixed, char)
                     eq_(apply_plan(garb, plan), char)
+            else:
+                print("%r <- %r" % (char, garble))
```
Comment (alin-luminoso) [inline tests/test_characters.py]: Is this addition intentional?

File: tests/test_synthetic.py [hunk 2]
```diff
@@ -67,3 +67,20 @@ def test_unknown_emoji():
...
+
+    # We believe enough in the future of this codepoint that we'll even
+    # recognize it with a mangled byte A0
+    emojibake = emojibake.replace('\xa0', ' ')
+
+    # Increment the first byte to get a very similar test case, but a
+    # codepoint that will definitely not exist anytime soon. In this case,
+    # we consider the existing text, "ñŸ¥\xa0", to be more probable.
+    not_emoji = "\U0005f960 I see mojibake in your present"
+    eq_(fix_encoding(not_emoji), not_emoji)
```
Comment (alin-luminoso) [inline tests/test_synthetic.py]: Am I misunderstanding this test?  Wouldn't it want to apply to `not_emoji.encode().decode()`?

**Comment by alin-luminoso** [comment]
Other than as commented, this looks fine.  I'll want to run benchmarks just to be sure (but don't expect to turn anything up).

File: tests/test_synthetic.py [hunk 3]
```diff
@@ -67,3 +67,20 @@ def test_unknown_emoji():
...
+    # The range we accept as emoji has gotten larger. Let's make sure we can
+    # decode the futuristic emoji U+1F960, which will probably be a picture of
+    # a fortune cookie in Unicode 10.0:
+    emoji_text = "\U0001f960 I see emoji in your future"
+    emojibake = emoji_text.encode('utf-8').decode('windows-1252')
+    eq_(fix_encoding(emojibake), emoji_text)
+
+    # We believe enough in the future of this codepoint that we'll even
+    # recognize it with a mangled byte A0
+    emojibake = emojibake.replace('\xa0', ' ')
```
Comment (alin-luminoso) [inline tests/test_synthetic.py]: It doesn't seem to do anything with this line after the replacement.

File: tests/test_synthetic.py (hunk 2 above)
Comment (rspeer) [inline tests/test_synthetic.py]: Oh right, yeah.

File: tests/test_characters.py (hunk 1 above)
Comment (rspeer) [inline tests/test_characters.py]: Nope.

File: tests/test_characters.py [hunk 4]
```diff
@@ -33,6 +35,16 @@ def test_possible_encoding():
...
 
 
+def test_byte_order_mark():
+    eq_(fix_encoding('ï»¿'), '\ufeff')
+
+
+def test_emoji_variation_selector():
+    # The hearts here are explicitly marked as emoji using the variation
+    # selector U+FE0F. This is not weird.
+    eq_(sequence_weirdness('❤️❤️❤️❤️❤️❤️❤️❤️❤️'), 0)
```
Comment (alin-luminoso) [inline tests/test_characters.py]: Would the test still work if the string were spelled, e.g., `'❤ \ufe0f' * 9`?  Might be plainer to the casual observer (even without the comment).

File: tests/test_synthetic.py [hunk 5]
```diff
@@ -67,3 +67,21 @@ def test_unknown_emoji():
...
+    eq_(fix_encoding(emojibake), emoji_text)
+
+    # Increment the first byte to get a very similar test case, but a
+    # codepoint that will definitely not exist anytime soon. In this case,
+    # we consider the existing text, "ñŸ¥\xa0", to be more probable.
+    not_emoji = "\U0005f960 I see mojibake in your present".encode('utf-8').decode('windows-1252')
+    eq_(fix_encoding(not_emoji), not_emoji)
```
Comment (alin-luminoso) [inline tests/test_synthetic.py]: Of course, now that I look at it -- is this actually functionally different from the last "dos burritos" test above?

Comment (rspeer) [inline tests/test_synthetic.py]: The difference is that the burrito is a Unicode 9 emoji that's already in the range we considered typical for emoji. The fortune cookie is a Unicode 10 emoji with a codepoint higher than we expected. Previous versions of ftfy would see U+1F960 and say "whoa, that's a high unassigned codepoint, I don't know what that is and that's scary", just like they would for U+5F960.

Maybe this makes the burrito test kind of obsolete, except the burrito test uses different mojibake, so the heuristics of why it's a close decision whether to fix it are different.

Comment (alin-luminoso) [inline tests/test_synthetic.py]: Oh, I meant specifically the `\U0003f32f` part.  (It doesn't seem like a big deal anyway, though.)

Comment (alin-luminoso) [inline tests/test_synthetic.py]: ...oh, but now I notice that that is windows-1251 and this is windows-1252.
